src/EditorFeatures/Test/Snippets/RoslynLSPSnippetConvertTests.cs

```diff
@@ -0,0 +1,243 @@
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.UnitTests.Workspaces;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Snippets;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Editor.UnitTests.Snippets
+{
+    [UseExportProvider]
+    public class RoslynLSPSnippetConvertTests
+    {
+        #region Edgecase extend TextChange tests
+
+        [Fact, Trait(Traits.Feature, Traits.Features.RoslynLSPSnippetConverter)]
+        public Task TestExtendSnippetTextChangeForwardsForCaret()
+        {
+            var markup =
+@"[|if ({|placeholder:true|})
+{
+}|] $$";
+
+            var expectedLSPSnippet =
+@"if (${1:true})
+{
+} $0";
+
+            return TestAsync(markup, expectedLSPSnippet);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.RoslynLSPSnippetConverter)]
+        public Task TestExtendSnippetTextChangeBackwardsForCaret()
+        {
+            var markup =
+@"$$ [|if ({|placeholder:true|})
+{
+}|]";
+
+            var expectedLSPSnippet =
+@"$0 if (${1:true})
+{
+}";
+
+            return TestAsync(markup, expectedLSPSnippet);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.RoslynLSPSnippetConverter)]
+        public Task TestExtendSnippetTextChangeForwardsForPlaceholder()
+        {
+            var markup =
+@"[|if (true)
+{$$
+}|] {|placeholder:test|}";
+
+            var expectedLSPSnippet =
+@"if (true)
+{$0
+} ${1:test}";
+
+            return TestAsync(markup, expectedLSPSnippet);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.RoslynLSPSnippetConverter)]
+        public Task TestExtendSnippetTextChangeBackwardsForPlaceholder()
+        {
+            var markup =
+@"{|placeholder:test|} [|if (true)
+{$$
+}|]";
+
+            var expectedLSPSnippet =
+@"${1:test} if (true)
+{$0
+}";
+
+            return TestAsync(markup, expectedLSPSnippet);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.RoslynLSPSnippetConverter)]
+        public Task TestExtendSnippetTextChangeForwardsForPlaceholderThenCaret()
+        {
+            var markup =
+@"[|if (true)
+{
+}|] {|placeholder:test|} $$";
+
+            var expectedLSPSnippet =
+@"if (true)
+{
+} ${1:test} $0";
+
+            return TestAsync(markup, expectedLSPSnippet);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.RoslynLSPSnippetConverter)]
+        public Task TestExtendSnippetTextChangeForwardsForCaretThenPlaceholder()
+        {
+            var markup =
+@"[|if (true)
+{
+}|] $$ {|placeholder:test|}";
+
+            var expectedLSPSnippet =
+@"if (true)
+{
+} $0 ${1:test}";
+
+            return TestAsync(markup, expectedLSPSnippet);
+        }
```

Comment:
nice.  i love hte above two tests.  do we have the same concept except where the caret/placeholders come before the text?

Also, consider some tests where there is code *both* before/after the [| |] region.